File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] heights) {
        int n = heights.length;
        int[] prefix = new int[n];
        int[] suffix = new int[n];
        prefix[0] = heights[0];
        suffix[n-1] = heights[n-1];
        for(int i =1;i<n;i++){
            prefix[i] = prefix[i-1]>heights[i] ? prefix[i-1] : heights[i];
        }
        for(int i=n-2;i>=0;i--){
            suffix[i] = suffix[i+1]>heights[i] ? suffix[i+1] : heights[i];
        }
        int ans = 0;
        for(int i=0;i<n;i++){
            ans += (Math.min(prefix[i],suffix[i])-heights[i]);
        }
        return ans;
    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self,heights):
        n = len(heights)
        if n == 0:
            return 0
        
        prefix = [0] * n
        suffix = [0] * n
        prefix[0] = heights[0]
        suffix[n - 1] = heights[n - 1]
        
        for i in range(1, n):
            prefix[i] = max(prefix[i - 1], heights[i])
        
        for i in range(n - 2, -1, -1):
            suffix[i] = max(suffix[i + 1], heights[i])
        
        ans = 0
        for i in range(n):
            ans += min(prefix[i], suffix[i]) - heights[i]
        
        return ans